File: app/lib/bots/dorothy_dca.dart
import 'dart:math' as math;
import '../core/models/candle.dart';
import '../core/models/order.dart';
import '../core/models/portfolio.dart';
import 'bot_base.dart';

/// DCA ladder strategy: buy on first opportunity, then buy more on dips
/// from the last anchor (up to maxPositions). Sell all at a profit target.
/// Full liquidation if price drops below avg cost * (1 - stopLossPct).
class DorothyDCA extends BotBase {
  final double profitFactorPct;
  final double marginDropPct;
  final int maxPositions;
  final double stopLossPct;
  final double quoteOrderQty;

  @override
  final String id;
  @override
  final String name;

  double? _lastBuyPrice;
  double? _sellTarget;

  DorothyDCA({
    this.id = 'dorothy_dca',
    this.name = 'Dorothy DCA',
    this.profitFactorPct = 5.0,
    this.marginDropPct = 0.4,
    this.maxPositions = 3,
    this.stopLossPct = 15.0,
    this.quoteOrderQty = 7.0,
  });

  @override
  Map<String, dynamic> get params => {
        'profitFactorPct': profitFactorPct,
        'marginDropPct': marginDropPct,
        'maxPositions': maxPositions,
        'stopLossPct': stopLossPct,
        'quoteOrderQty': quoteOrderQty,
      };

  @override
  List<BotParamSpec> paramSpec() => const [
        BotParamSpec(name: 'profitFactorPct', label: 'Profit Target %', type: BotParamType.percent, defaultValue: 5.0, min: 0.5, max: 50.0, step: 0.5),
        BotParamSpec(name: 'marginDropPct', label: 'Extra Drop %', type: BotParamType.percent, defaultValue: 0.4, min: 0.1, max: 20.0, step: 0.1),
        BotParamSpec(name: 'maxPositions', label: 'Max Positions', type: BotParamType.intParam, defaultValue: 3, min: 1, max: 10, step: 1),
        BotParamSpec(name: 'stopLossPct', label: 'Stop Loss %', type: BotParamType.percent, defaultValue: 15.0, min: 0.0, max: 50.0, step: 1.0),
        BotParamSpec(name: 'quoteOrderQty', label: 'USDT per order', type: BotParamType.doubleParam, defaultValue: 7.0, min: 1.0, max: 1000.0, step: 1.0),
      ];

  @override
  List<OrderRequest> onCandle(Candle candle, Portfolio portfolio) {
    final price = candle.close;
    final myPositions = portfolio.positionsForBot(id);
    final nOpen = myPositions.length;
    final orders = <OrderRequest>[];

    // Stop loss: liquidate all if drawdown vs avg cost is too deep.
    if (nOpen > 0 && stopLossPct > 0) {
      final totalQty = myPositions.fold(0.0, (s, p) => s + p.qty);
      final totalCost =
          myPositions.fold(0.0, (s, p) => s + p.entryPrice * p.qty);
      final avgCost = totalCost / totalQty;
      if (price < avgCost * (1 - stopLossPct / 100)) {
        _lastBuyPrice = null;
        _sellTarget = null;
        if (totalQty > 0) {
          orders.add(OrderRequest.closeLong(botId: id, qty: totalQty));
        }
        return orders;
      }
    }

    // Take profit: sell all when target hit.
    if (nOpen > 0 && _sellTarget != null && price >= _sellTarget!) {
      final qty = maxSellQty(portfolio);
      _lastBuyPrice = null;
      _sellTarget = null;
      if (qty > 0) orders.add(OrderRequest.closeLong(botId: id, qty: qty));
      return orders;
    }

    // Entry: first buy or DCA dip
    if (nOpen < maxPositions) {
      bool shouldBuy = false;
      if (_lastBuyPrice == null) {
        shouldBuy = true;
      } else {
        final dropThreshold = _lastBuyPrice! *
            (1 - (profitFactorPct + marginDropPct) / 100);
        if (price <= dropThreshold) shouldBuy = true;
      }
      if (shouldBuy) {
        final actualUsdt = math.min(portfolio.cash, quoteOrderQty);
        final qty = price > 0 ? actualUsdt / price : 0.0;
        if (qty > 0) {
          orders.add(OrderRequest.openLong(botId: id, qty: qty));
          _lastBuyPrice = price;
          _sellTarget = price * (1 + profitFactorPct / 100);
        }
      }
    }

    return orders;
  }

  @override
  BotState? state() => BotState(
        phase: _lastBuyPrice == null ? 'idle' : 'live',
        decision: _sellTarget != null ? 'awaiting_tp' : 'scanning',
        indicators: {
          'last_buy': ?_lastBuyPrice,
          'sell_target': ?_sellTarget,
        },
      );
}

File: backtester/bots/dorothy_dca.py
"""DorothyDCA — ladder/DCA strategy adapted from Pecunator Dorothy v7.

Strategy:
  - On first candle with no position: BUY a slot.
  - After a buy, set a virtual SELL target at buy_price * (1 + profit_factor).
  - SELL (take profit) when price >= sell_target.
  - Re-BUY when price drops from last_buy_price by (profit_factor + margin_drop_factor),
    up to max_positions open simultaneously.
  - Stop-loss: if price drops below avg_cost * (1 - stop_loss_pct), liquidate all.
"""

from typing import Any

from backtester.bots.bot_base import BotBase
from backtester.core.engine import Candle, Portfolio


class DorothyDCA(BotBase):
    """DCA ladder strategy inspired by Pecunator Dorothy.

    Buys on dips from the last buy anchor and sells each rung at a
    fixed profit target. Multiple concurrent positions are allowed up
    to max_positions.
    """

    def __init__(
        self,
        profit_factor: float = 0.05,
        margin_drop_factor: float = 0.004,
        max_positions: int = 3,
        stop_loss_pct: float = 0.15,
        quote_order_qty: float = 7.0,
    ) -> None:
        self.profit_factor = profit_factor
        self.margin_drop_factor = margin_drop_factor
        self.max_positions = max_positions
        self.stop_loss_pct = stop_loss_pct
        self.quote_order_qty = quote_order_qty

        self._last_buy_price: float | None = None
        self._sell_target: float | None = None
        self._warmup_done: bool = False

    @classmethod
    def param_spec(cls) -> dict[str, dict[str, Any]]:
        return {
            "profit_factor": {
                "type": "float",
                "default": 0.05,
                "min": 0.005,
                "max": 0.5,
                "step": 0.005,
            },
            "margin_drop_factor": {
                "type": "float",
                "default": 0.004,
                "min": 0.001,
                "max": 0.2,
                "step": 0.001,
            },
            "max_positions": {
                "type": "int",
                "default": 3,
                "min": 1,
                "max": 10,
                "step": 1,
            },
            "stop_loss_pct": {
                "type": "float",
                "default": 0.15,
                "min": 0.0,
                "max": 0.5,
                "step": 0.01,
            },
            "quote_order_qty": {
                "type": "float",
                "default": 7.0,
                "min": 1.0,
                "max": 1000.0,
                "step": 1.0,
            },
        }

    def on_candle(self, candle: Candle, portfolio: Portfolio) -> list[dict[str, Any]]:
        price = float(candle.close)
        orders: list[dict[str, Any]] = []
        n_open = len(portfolio.positions)

        # ── Stop-loss: liquidate all if price too far below avg cost ──
        # stop_loss_pct <= 0 disables the protection entirely.
        if n_open > 0 and self.stop_loss_pct > 0:
            avg_cost = float(
                sum(p.entry_price * p.qty for p in portfolio.positions)
                / sum(p.qty for p in portfolio.positions)
            )
            if price < avg_cost * (1 - self.stop_loss_pct):
                qty = self.max_sell_qty(portfolio)
                if qty > 0:
                    orders.append(
                        {"side": "SELL", "qty": float(qty), "reason": "STOP_LOSS"}
                    )
                self._last_buy_price = None
                self._sell_target = None
                return orders

        # ── Take profit: sell all when price reaches target ───────────
        if n_open > 0 and self._sell_target is not None:
            if price >= self._sell_target:
                qty = self.max_sell_qty(portfolio)
                if qty > 0:
                    orders.append(
                        {"side": "SELL", "qty": float(qty), "reason": "TAKE_PROFIT"}
                    )
                self._last_buy_price = None
                self._sell_target = None
                return orders

        # ── Entry: first buy or DCA dip ───────────────────────────────
        if n_open < self.max_positions:
            should_buy = False

            if self._last_buy_price is None:
                # No position yet — buy on first opportunity
                should_buy = True
            else:
                # DCA: buy when price drops enough from last anchor
                drop_threshold = self._last_buy_price * (
                    1 - (self.profit_factor + self.margin_drop_factor)
                )
                if price <= drop_threshold:
                    should_buy = True

            if should_buy:
                actual_usdt = min(float(portfolio.cash), self.quote_order_qty)
                qty = actual_usdt / price if price > 0 else 0.0
                if qty > 0:
                    orders.append(
                        {"side": "BUY", "qty": float(qty), "reason": "DCA_BUY"}
                    )
                    self._last_buy_price = price
                    # Sell target is always relative to latest buy price
                    self._sell_target = price * (1 + self.profit_factor)

        return orders
